packages/hagrid/hagrid/cli.py

```diff
@@ -351,40 +353,164 @@ def launch(args: TypeTuple[str], **kwargs: Any) -> None:
+        command = cmds[list(cmds.keys())[0]][0]  # type: ignore
+        match_port = re.search("HTTP_PORT=[0-9]{1,5}", command)
+
+        if host_term.host == "docker" and match_port:
+            rich.get_console().print(
+                "\n[bold green]⠋[bold blue] Checking  Node API [/bold blue]\t"
+            )
+            port = match_port.group().replace("HTTP_PORT=", "")
+            check_status("localhost" + ":" + port)
     except Exception as e:
         print(f"Error: {e}\n\n")
         return
 
 
-def execute_commands(
-    cmds: TypeList,
-    dry_run: bool = False,
-    silent: bool = False,
-    from_rendered_dir: bool = False,
-) -> None:
-    """Execute the launch commands and display their status in realtime.
+def check_errors(line: str, pid: int, cmd_name: str, progress_bar: Progress) -> None:
+    task = progress_bar.tasks[0]
+    if "Error response from daemon: " in line:
+        if progress_bar:
+            progress_bar.update(
+                0,
+                description=f"❌ [bold red]{cmd_name}[/bold red] [{task.completed} / {task.total}]",
+                refresh=True,
+            )
+            progress_bar.update(0, visible=False)
+            progress_bar.console.clear_live()
+            progress_bar.console.quiet = True
+            progress_bar.stop()
+            console = rich.get_console()
+            progress_bar.console.quiet = False
+            console.print(f"\n\n [red] ERROR [/red]: [bold]{line}[/bold]\n")
+        os.killpg(os.getpgid(pid), sys_signal.SIGTERM)
+        raise Exception
+
+
+def check_pulling(line: str, cmd_name: str, progress_bar: Progress) -> None:
+    task = progress_bar.tasks[0]
+    if "Pulling" in line and "fs layer" not in line:
+        progress_bar.update(
+            0,
+            description=f"⌛ [bold]{cmd_name} [{task.completed} / {task.total+1}]",
+            total=task.total + 1,
+            refresh=True,
+        )
+    if "Pulled" in line:
+        progress_bar.update(
+            0,
+            description=f"⌛ [bold]{cmd_name} [{task.completed + 1} / {task.total}]",
+            completed=task.completed + 1,
+            refresh=True,
+        )
+        if progress_bar.finished:
+            progress_bar.update(
+                0,
+                description=f"✅ [bold green]{cmd_name} [{task.completed} / {task.total}]",
+                refresh=True,
+            )
 
-    Args:
-        cmds (list): list of commands to be executed
-        dry_run (bool, optional): If `True` only displays cmds to be executed. Defaults to False.
-    """
-    process_list: TypeList = []
-    console = rich.get_console()
 
-    username, password = (
-        extract_username_and_pass(cmds[0]) if len(cmds) > 0 else ("-", "-")
-    )
-    # display VM credentials
-    console.print(generate_user_table(username=username, password=password))
+def check_building(line: str, cmd_name: str, progress_bar: Progress) -> None:
```

Comment:
We need to improve how to track docker logs in the building step since it's difficult to know beforehand dynamically how many steps each image has to build _(we usually know this when the images are already built)_. For now, I set 55 as the total amount of steps but it's not very reliable since we  can change it very often

Reply:
Ok, that's an awesome suggestion, I'll definitely try this. Thank you @shubham3121

Reply:
Solved by : [eba6a64](https://github.com/OpenMined/PySyft/pull/6900/commits/eba6a649322f42e2c931fb73dd4fa2d25a58a18f)

Reply:
I found a different approach that is also reliable and make and also counts images built instead of generic steps from the building of all the images.

This makes more sense if we consider that pull also counts the number of images being pulled, not the steps to pull them, and launch counts the number of containers being launched.